packages/project/test/merge/map-uuids.test.ts

```diff
@@ -71,3 +71,251 @@ test('one connecting node missing', (t) => {
+  // set expression to b and x
+  const expression = 'fn(s => s)';
+  source.set('b', { expression });
+  target.set('x', { expression });
+
+  const result = mapUUIDs(source.workflow, target.workflow);
+  t.deepEqual(result.nodes, {
+    ['a']: target.getUUID('a'),
+    ['b']: target.getUUID('x'),
+    ['x']: null,
+  });
+  t.deepEqual(result.edges, {
+    ['a-b']: target.getUUID('a-x'),
+    ['a-x']: null,
+  });
+});
+
+test('edge removed between existing nodes', (t) => {
+  const source = generateWorkflow(['a-b']);
+  const target = generateWorkflow(['a-b', 'a-c']);
+
+  const result = mapUUIDs(source.workflow, target.workflow);
+  t.deepEqual(result.nodes, {
+    ['a']: target.getUUID('a'),
+    ['b']: target.getUUID('b'),
+    ['c']: null,
+  });
+  t.deepEqual(result.edges, {
+    ['a-b']: target.getUUID('a-b'),
+    ['a-c']: null,
+  });
+});
+
+test('new edge added', (t) => {
+  const source = generateWorkflow(['a-b', 'a-c']);
+  const target = generateWorkflow(['a-b']);
+
+  const result = mapUUIDs(source.workflow, target.workflow);
+  t.deepEqual(result.nodes, {
+    ['a']: target.getUUID('a'),
+    ['b']: target.getUUID('b'),
+    ['c']: true,
+  });
+  t.deepEqual(result.edges, {
+    ['a-b']: target.getUUID('a-b'),
+    ['a-c']: true,
+  });
+});
+
+test('multiple steps and edges added and removed', (t) => {
+  const source = generateWorkflow(['a-b', 'b-c', 'c-d']);
+  const target = generateWorkflow(['a-x', 'x-b', 'b-y', 'y-c', 'c-d', 'd-z']);
+
+  const result = mapUUIDs(source.workflow, target.workflow);
+  t.deepEqual(result.nodes, {
+    ['a']: target.getUUID('a'),
+    ['b']: target.getUUID('b'),
+    ['c']: target.getUUID('c'),
+    ['d']: target.getUUID('d'),
+    ['x']: null,
+    ['y']: null,
+    ['z']: null,
+  });
+  t.deepEqual(result.edges, {
+    ['a-b']: true,
+    ['b-c']: true,
+    ['c-d']: target.getUUID('c-d'),
+    ['a-x']: null,
+    ['x-b']: null,
+    ['b-y']: null,
+    ['y-c']: null,
+    ['d-z']: null,
+  });
+});
+
+test('steps removed from a workflow', (t) => {
+  const source = generateWorkflow(['a-b', 'a-c', 'b-d', 'c-d']);
+  const target = generateWorkflow(['a-x', 'x-b', 'a-c', 'b-d', 'c-d', 'd-e']);
+
+  const result = mapUUIDs(source.workflow, target.workflow);
+  t.deepEqual(result.nodes, {
+    ['a']: target.getUUID('a'),
+    ['b']: target.getUUID('b'),
+    ['c']: target.getUUID('c'),
+    ['d']: target.getUUID('d'),
+    ['e']: null,
+    ['x']: null,
+  });
+  t.deepEqual(result.edges, {
+    ['a-b']: true,
+    ['a-c']: target.getUUID('a-c'),
+    ['b-d']: target.getUUID('b-d'),
+    ['c-d']: target.getUUID('c-d'),
+    ['a-x']: null,
+    ['x-b']: null,
+    ['d-e']: null,
+  });
+});
+
+// NEW TESTS
+
+// narrowing by parent
+// id changed but parent stayed the same
+test('parent stayed same but id changed', (t) => {
+  const source = generateWorkflow(['a-b']);
+  const target = generateWorkflow(['a-x']);
+
+  const result = mapUUIDs(source.workflow, target.workflow);
```

Comment:
Wait, shouldn't we be passing the `Workflow` class directly into mapUUIDs? `Workflow` should have utility functions that can help us

Reply:
that's correct. resolved